app/javascript/src/NewService/components/NewServiceForm.jsx

```diff
@@ -0,0 +1,150 @@
+    .catch(error => {
+      setError({hasError: true, errorMsg: error.message})
+    })
+}
+
+const callEffect = (servicesRequested, selectedNamespace, startServicesFetch) => {
+  if (servicesRequested && selectedNamespace) {
+    startServicesFetch(`${BASE_PATH}/namespaces/${selectedNamespace}/services.json`)
+  }
+}
+
+const renderForm = (sourceFormProps, serviceDiscoveryProps, serviceNewProps) =>
+  <React.Fragment>
+    <h1>New API</h1>
+    <ServiceSourceForm {...sourceFormProps}/>
+    <FormTemplate {...serviceDiscoveryProps}/>
+    <FormTemplate {...serviceNewProps}/>
+  </React.Fragment>
+
+const onNamespacesFetch = async (
+  setNamespaces, setServicesRequested, settingSelectedNamespace, setError
+): Promise<void> => {
+  setNamespaces([])
+  const namespacesData = await fetchData(PROJECTS_PATH, setError)
+  if (namespacesData) {
+    setNamespaces(namespacesData)
+    setServicesRequested(true)
+    settingSelectedNamespace(namespacesData[0])
+  }
+}
+
+const onServicesFetch = async (
+  servicesUrl, setServices, setServicesRequested, setError
+): Promise<void> => {
+  setServices([])
+  const servicesData = await fetchData(servicesUrl, setError)
+  if (servicesData) {
+    setServices(servicesData)
+    setServicesRequested(false)
+  }
+}
+
+const getChangeNamespaces = (event, setSelectedNamespace, setServicesRequested) => {
+  setSelectedNamespace(event.currentTarget.value)
+  setServicesRequested(true)
+}
+
+const getHandleFormsVisibility = (event, setServDescVisible, setServNewVisible) => {
+  if (event.currentTarget.value === 'discover') {
+    setServDescVisible(true)
+    setServNewVisible(false)
+  } else {
+    setServDescVisible(false)
+    setServNewVisible(true)
+  }
+}
+
+const NewServiceForm = (props: Props) => {
+  const [isServDescVisible, setServDescVisible] = useState(false)
+  const [isServNewVisible, setServNewVisible] = useState(true)
+  const [namespaces, setNamespaces] = useState([])
+  const [services, setServices] = useState([])
+  const [selectedNamespace, setSelectedNamespace] = useState(undefined)
+  const [servicesRequested, setServicesRequested] = useState(false)
+  const [error, setError] = useState({hasError: false, errorMsg: ''})
+
+  const settingSelectedNamespace = (namespace: string) =>
+    setSelectedNamespace(namespace)
+  const onChangeNamespaces = (event) =>
+    getChangeNamespaces(event, setSelectedNamespace, setServicesRequested)
+  const onStartNamespacesFetch = () =>
+    onNamespacesFetch(setNamespaces, setServicesRequested, settingSelectedNamespace, setError)
+  const startServicesFetch = (servicesUrl) =>
+    onServicesFetch(servicesUrl, setServices, setServicesRequested, setError)
+  const onHandleFormsVisibility = (event) =>
+    getHandleFormsVisibility(event, setServDescVisible, setServNewVisible)
+
+  useEffect(
+    () => callEffect(servicesRequested, selectedNamespace, startServicesFetch),
```

Comment:
'callEffect` is a bit redundant, either define a custom hook or declare useEffect explicitly.